### ISSUE TITLE
Maximum update depth exceeded when selecting a text

### ISSUE THREAD

**Issue by nomadus** [outline/rich-markdown-editor#39]
I am having an issue, there is a similar closed issue, but the stack trace is different. I installed the latest version and tried to play with an example. It happens when I select a text. When I select first time it is OK, the error happens if I select the second time.

```invariant.js:42 Uncaught Error: Maximum update depth exceeded. 
This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    at invariant (invariant.js:42)
    at scheduleWork$1 (react-dom.development.js:16222)
    at Object.enqueueSetState (react-dom.development.js:11300)
    at RichMarkdownEditor.Component.setState (react.development.js:270)
    at handleChange (index.js:204)
    at Editor.onChange (slate-react.js:4183)
    at Editor.change (slate-react.js:4154)
    at Editor.onEvent (slate-react.js:4170)
    at Object._this.(:8080/anonymous function) [as onSelect] (webpack:///./node_modules/slate-react/lib/slate-react.js?:3921:23)
    at HTMLDocument.eval (slate-react.js:1960)``

**Comment by tommoor** [comment]
Are you able to provide the exact contents of the editor? If running the example with `yarn start` you can access this from the console with `localStorage.saved`.

I'm currently unable to reproduce locally so it's likely the editor content is important.

**Comment by nomadus** [comment]
Hi, were you able to reproduce the issue?

**Comment by tommoor** [comment]
@nomadus I've been unable to reproduce with your content above :( I don't doubt there is an issue though, could you post a screen capture perhaps?

**Comment by nomadus** [comment]
I can reproduce it at will, I have recorded the steps. To reproduce you need to format the strings and then try to double click on each strings, at some point the error is thrown.
Here is the recording
https://drive.google.com/open?id=16j5EqD5aMhaoqD-AhJzlkTt5jPa6fgUP

**Comment by nomadus** [comment]
Just in case, copying the stack trace, as I said earlier, this happens when I keep selecting/deselecting a random text

Uncaught Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    at invariant (invariant.js:42)
    at scheduleWork$1 (react-dom.development.js:16222)
    at Object.enqueueSetState (react-dom.development.js:11300)
    at RichMarkdownEditor.Component.setState (react.development.js:270)
    at handleChange (index.js:204)
    at Editor.onChange (slate-react.js:4183)
    at Editor.change (slate-react.js:4154)
    at Editor.onEvent (slate-react.js:4170)
    at Object._this.(:55463/anonymous function) [as onSelect] (webpack:///./node_modules/slate-react/lib/slate-react.js?:3921:23)
    at HTMLDocument.eval (slate-react.js:1960)

**Comment by nomadus** [comment]
I think, I find a shorter way to reproduce the issue. Please see the video below, it happens when you click on a menu button (in my case H1), notice that the mouse pointer becomes a hand, whenever this happens, just move to a blank space and double click. 

https://drive.google.com/open?id=1pyDHKmpGQ7UED4W1fI7xGbNTN7emuQ6_

**Comment by nomadus** [comment]
Update
In the RichMarkdownEditor, I commented a line of code, it reads as below now, when I did so and reproduced the issue, it comes with  a more detailed error, the issue, I believe is in Toolbar component.

```
  setEditorRef = (ref: Editor) => {
    this.editor = ref;
    // Force re-render to show ToC (<Content />)
    //this.setState({ editorLoaded: true });
  };
```

invariant.js:42 Uncaught Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
    at invariant (invariant.js:42)
    at scheduleWork$1 (react-dom.development.js:16222)
    at Object.enqueueSetState (react-dom.development.js:11300)
    at StyledComponent.Component.setState (react.development.js:270)
    at StyledComponent.componentWillReceiveProps (styled-components.browser.cjs.js:1819)
    at callComponentWillReceiveProps (react-dom.development.js:11528)
    at updateClassInstance (react-dom.development.js:11720)
    at updateClassComponent (react-dom.development.js:13154)
    at beginWork (react-dom.development.js:13825)
    at performUnitOfWork (react-dom.development.js:15864)
invariant @ invariant.js:42
scheduleWork$1 @ react-dom.development.js:16222
enqueueSetState @ react-dom.development.js:11300
Component.setState @ react.development.js:270
componentWillReceiveProps @ styled-components.browser.cjs.js:1819
callComponentWillReceiveProps @ react-dom.development.js:11528
updateClassInstance @ react-dom.development.js:11720
updateClassComponent @ react-dom.development.js:13154
beginWork @ react-dom.development.js:13825
performUnitOfWork @ react-dom.development.js:15864
workLoop @ react-dom.development.js:15903
callCallback @ react-dom.development.js:100
invokeGuardedCallbackDev @ react-dom.development.js:138
invokeGuardedCallback @ react-dom.development.js:187
replayUnitOfWork @ react-dom.development.js:15311
renderRoot @ react-dom.development.js:15963
performWorkOnRoot @ react-dom.development.js:16561
performWork @ react-dom.development.js:16483
performSyncWork @ react-dom.development.js:16455
requestWork @ react-dom.development.js:16355
scheduleWork$1 @ react-dom.development.js:16219
enqueueSetState @ react-dom.development.js:11300
Component.setState @ react.development.js:270
handleChange @ index.js:200
onChange @ slate-react.js:4183
change @ slate-react.js:4154
onEvent @ slate-react.js:4170
_this.(anonymous function) @ slate-react.js:3921
(anonymous) @ slate-react.js:1960
invokeFunc @ debounce.js:95
leadingEdge @ debounce.js:105
debounced @ debounce.js:172
react-dom.development.js:14227 The above error occurred in the <Toolbar__Menu> component:
    in Toolbar__Menu (created by Toolbar)
    in Portal (created by Toolbar)
    in Toolbar (created by RichMarkdownEditor)
    in ThemeProvider (created by RichMarkdownEditor)
    in div (created by Flex__Container)
    in Flex__Container (created by Flex)
    in Flex (created by RichMarkdownEditor)
    in RichMarkdownEditor (created by Example)
    in div (created by Example)
    in Example

**Comment by tommoor** [comment]
Thanks for the extra details – are you able to make the videos public? They're 404ing for me right now. Either way there should be enough details here to try and fix

**Comment by nomadus** [comment]
Hi,
Just in case adding the links once more, although google says that anyone with the link could view,
https://drive.google.com/file/d/16j5EqD5aMhaoqD-AhJzlkTt5jPa6fgUP/view?usp=sharing
https://drive.google.com/file/d/1pyDHKmpGQ7UED4W1fI7xGbNTN7emuQ6_/view?usp=sharing

**Comment by ShiroYacha** [comment]
Hey guys,

Actually I had this issue for me in my app too. I wasn't able to reproduce this in the example here so I thought it might be a side-effect I have introduced. 

However, I was able to find a "workaround" so that this never happens again: I debounced the `update()` method for the toolbar. 

You can check it out at https://github.com/ShiroYacha/rich-markdown-editor/tree/debounce-toolbar or specifically here: https://github.com/ShiroYacha/rich-markdown-editor/blob/dbe39957504b1d26f5248b81741bba81dcc4e740/src/components/Toolbar/index.js

Although I am not sure this is the clean way... but I guess it could be useful as a temporary fix for you.

Best,

**Comment by tommoor** [comment]
Hey folks, sorry – I've been on holiday the last week. I think debouncing the update could be a great solution, let me check that out.

**Comment by tommoor** [comment]
@nomadus I just pushed a potential fix to master along with another fix. Would you be able to try it out if you get a chance before I publish an official release? 

I've tested there are no negative effects to the change but I've never quite been able to reproduce the issue you were seeing locally.

**Comment by tommoor** [comment]
Fixed by https://github.com/outline/rich-markdown-editor/commit/b32fe973936eb814ff3f42ba0a046ed52ef9912f

**Comment by Zoe314** [comment]
Hello, I have encountered the same problem. Can you solve it? What is the method? I saw the tips of the official website, but did not see the solution, tell me the trouble! Thank you very much! God bless you!
My error is: react-dom.production.min.js:179 Invariant Violation: Minified React error #185; Visit

**Comment by gigiskarlett** [comment]
@tommoor Hi, do you have any idea why clearing browser history and cache fixes the Minified React error #185 temporarily?

**Comment by tommoor** [comment]
I don't, I haven't seen this error since it was fixed last year
